src/firmware_logger_device.cpp

```diff
@@ -44,91 +66,151 @@ namespace librealsense
-            fw_logs::fw_logs_binary_data binary_data{ resultsForOneLog };
-            _flash_logs.push(binary_data);
-            beginOfLogIterator = endOfLogIterator;
-        }
+    command firmware_logger_device::get_update_command()
+    {
+        if( ! _parser )
+            throw librealsense::wrong_api_call_sequence_exception( "FW log parser is not initialized" );
 
-        _flash_logs_initialized = true;
+        command update_command = _parser->get_update_command();
+        update_command.cmd = _fw_logs_command.cmd; // Opcode comes from the device, may be different between devices
+
+        return update_command;
     }
 
-    bool firmware_logger_device::get_flash_log(fw_logs::fw_logs_binary_data& binary_data)
+    bool firmware_logger_device::init_parser( std::string xml_content )
     {
-        bool result = false;
-        if (!_flash_logs_initialized)
-        {
-            get_flash_logs_from_hw_monitor();
-        }
+        _parser = std::make_unique< fw_logs::fw_logs_parser >( xml_content );
 
-        if (!_flash_logs.empty())
+        return ( _parser != nullptr );
+    }
+
+    bool firmware_logger_device::parse_log( const fw_logs::fw_logs_binary_data * fw_log_msg,
+                                            fw_logs::fw_log_data * parsed_msg )
+    {
+        bool result = false;
+        if( _parser && parsed_msg && fw_log_msg )
         {
-            fw_logs::fw_logs_binary_data data;
-            data = _flash_logs.front();
-            _flash_logs.pop();
-            binary_data = data;
+            *parsed_msg = _parser->parse_fw_log( fw_log_msg );
             result = true;
         }
+
         return result;
     }
 
-    bool firmware_logger_device::init_parser(std::string xml_content)
+    legacy_firmware_logger_device::legacy_firmware_logger_device( std::shared_ptr< const device_info > const & dev_info,
+                                                                  std::shared_ptr< hw_monitor > hardware_monitor,
+                                                                  const command & fw_logs_command,
+                                                                  const command & flash_logs_command )
+        : firmware_logger_device( dev_info, hardware_monitor, fw_logs_command )
+        , _flash_logs_command( flash_logs_command )
+        , _flash_logs()
+        , _flash_logs_initialized( false )
     {
-        _parser = new fw_logs::fw_logs_parser(xml_content);
+    }
+
+    bool legacy_firmware_logger_device::init_parser( std::string xml_content )
+    {
+        _parser = std::make_unique< fw_logs::legacy_fw_logs_parser >( xml_content );
 
-        return (_parser != nullptr);
+        return ( _parser != nullptr );
     }
 
-    bool firmware_logger_device::parse_log(const fw_logs::fw_logs_binary_data* fw_log_msg,
-        fw_logs::fw_log_data* parsed_msg)
+    size_t legacy_firmware_logger_device::get_log_size( uint8_t * buff ) const
```

Comment:
Can this be inline method in the h file?

Reply:
virtual functions cannot be inlined. I prefer less code in the h file for clearer API.